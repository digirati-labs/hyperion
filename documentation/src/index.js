import './main.scss';
import '../../hyperion-logo.png';
import '../../atlas-logo.png';
import '../../vault-logo.png';
import '../../discover-logo.png';
import '../../react-vault-logo.png';

import 'simplebar';
import 'simplebar/dist/simplebar.css';
import '@canvas-panel/viewer-element/umd/@canvas-panel/viewer-element';
