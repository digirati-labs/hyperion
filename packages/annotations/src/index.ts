import { test, testFunction } from '@hyperion-framework/types';

console.log('hello world.');
console.log(test);
console.log(testFunction('test', '3'));
