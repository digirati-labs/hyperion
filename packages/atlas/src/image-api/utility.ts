import {ContentResource, IIIFExternalWebResource, ImageSize, ImageTile, Service} from '@hyperion-framework/types';
import {
  FixedSizeImageService,
  IIIF_1_IMAGE_LEVEL_0,
  IIIF_1_IMAGE_LEVEL_0_PROFILE,
  IIIF_1_IMAGE_LEVEL_1,
  IIIF_1_IMAGE_LEVEL_1_PROFILE,
  IIIF_1_IMAGE_LEVEL_2,
  IIIF_1_IMAGE_LEVEL_2_PROFILE,
  IIIF_2_IMAGE_LEVEL_0,
  IIIF_2_IMAGE_LEVEL_0_PROFILE,
  IIIF_2_IMAGE_LEVEL_1,
  IIIF_2_IMAGE_LEVEL_1_PROFILE,
  IIIF_2_IMAGE_LEVEL_2,
  IIIF_2_IMAGE_LEVEL_2_PROFILE,
  IIIF_3_IMAGE_LEVEL_0,
  IIIF_3_IMAGE_LEVEL_1,
  IIIF_3_IMAGE_LEVEL_2,
  ImageCandidate,
  STANFORD_IIIF_1_IMAGE_COMPLIANCE_0,
  STANFORD_IIIF_1_IMAGE_COMPLIANCE_1,
  STANFORD_IIIF_1_IMAGE_COMPLIANCE_2,
  STANFORD_IIIF_1_IMAGE_CONFORMANCE_0,
  STANFORD_IIIF_1_IMAGE_CONFORMANCE_1,
  STANFORD_IIIF_1_IMAGE_CONFORMANCE_2,
  STANFORD_IIIF_IMAGE_COMPLIANCE_0,
  STANFORD_IIIF_IMAGE_COMPLIANCE_1,
  STANFORD_IIIF_IMAGE_COMPLIANCE_2,
  STANFORD_IIIF_IMAGE_CONFORMANCE_0,
  STANFORD_IIIF_IMAGE_CONFORMANCE_1,
  STANFORD_IIIF_IMAGE_CONFORMANCE_2,
} from './types';

export function sizesMatch(sizesA: ImageSize[], sizesB: ImageSize[]): boolean {
  if (sizesA.length !== sizesB.length) {
    return false;
  }

  if (sizesA.length === 0 && sizesB.length === 0) {
    return true;
  }

  const len = sizesA.length;
  let matchOrder = true;
  for (let i = 0; i < len; i++) {
    const a = sizesA[i];
    const b = sizesB[i];
    if (a.width !== b.width || a.height !== b.height) {
      matchOrder = false;
      break;
    }
  }
  if (matchOrder) {
    return true;
  }

  let matching = 0;
  for (let a = 0; a < len; a++) {
    for (let b = 0; b < len; b++) {
      if (sizesA[a].width === sizesB[b].width && sizesA[a].height === sizesB[b].height) {
        matching++;
        break;
      }
    }
  }

  return matching === len;
}

export function fixedSizesFromScales(width: number, height: number, scales: number[]): ImageSize[] {
  const len = scales.length;
  const sizes: ImageSize[] = [];
  for (let i = 0; i < len; i++) {
    const scale = scales[i];
    sizes.push({
      width: Math.floor(width / scale),
      height: Math.floor(height / scale),
    });
  }
  return sizes;
}

export function extractFixedSizeScales(width: number, height: number, sizes: ImageSize[]): number[] {
  const len = sizes.length;
  const scales = [];
  for (let i = 0; i < len; i++) {
    const size = sizes[i];
    const w = size.width;
    scales.push(width / w);
  }
  return scales;
}

export function getFixedSizeFromImage(contentResource: ContentResource): ImageCandidate | null {
  if (typeof contentResource === 'string') {
    // Might not even be an image.
    return { type: 'unknown', id: contentResource };
  }

  if (contentResource.type !== 'Image') {
    return null;
  }

  const image = contentResource as IIIFExternalWebResource;

  if (!image.id) {
    return null;
  }

  if (image.id && image.width && image.height) {
    return {
      id: image.id,
      type: 'fixed',
      width: image.width,
      height: image.height,
    };
  }

  return { type: 'unknown', id: image.id };
}

export function getFixedSizesFromService(service: Service): FixedSizeImageService[] {
  return (service && service.sizes ? service.sizes : []).map(size => {
    return {
      id: service.id,
      type: 'fixed-service',
      height: size.height,
      width: size.width,
    };
  });
}

export function canonicalServiceUrl(serviceId: string) {
  return serviceId.endsWith('info.json')
    ? serviceId
    : serviceId.endsWith('/')
    ? `${serviceId}info.json`
    : `${serviceId}/info.json`;
}

export function getImageServerFromId(url: string): string {
  // Strip off the protocol + www
  const id = url.replace(/(https?:\/\/)?(www.)?/i, '');

  // Strip off the path.
  if (id.indexOf('/') !== -1) {
    return id.split('/')[0];
  }

  // Return the id.
  return id;
}

const level1Support = [
  STANFORD_IIIF_IMAGE_COMPLIANCE_1,
  STANFORD_IIIF_IMAGE_COMPLIANCE_2,
  STANFORD_IIIF_IMAGE_CONFORMANCE_1,
  STANFORD_IIIF_IMAGE_CONFORMANCE_2,
  STANFORD_IIIF_1_IMAGE_COMPLIANCE_1,
  STANFORD_IIIF_1_IMAGE_COMPLIANCE_2,
  STANFORD_IIIF_1_IMAGE_CONFORMANCE_1,
  STANFORD_IIIF_1_IMAGE_CONFORMANCE_2,
  IIIF_1_IMAGE_LEVEL_1,
  IIIF_1_IMAGE_LEVEL_1_PROFILE,
  IIIF_1_IMAGE_LEVEL_2,
  IIIF_1_IMAGE_LEVEL_2_PROFILE,
  IIIF_2_IMAGE_LEVEL_1,
  IIIF_2_IMAGE_LEVEL_1_PROFILE,
  IIIF_2_IMAGE_LEVEL_2,
  IIIF_2_IMAGE_LEVEL_2_PROFILE,
  IIIF_3_IMAGE_LEVEL_1,
  IIIF_3_IMAGE_LEVEL_2,
];

export function supportsCustomSizes(service: Service): boolean {
  if (!isImageService(service)) {
    return false;
  }

  const profiles = Array.isArray(service.profile) ? service.profile : [service.profile];

  for (const profile of profiles) {
    if (typeof profile === 'string') {
      if (level1Support.indexOf(profile) !== -1) {
        return true;
      }
    } else {
      const supports = profile.supports || [];
      if (
        supports.indexOf('regionByPx') !== -1 &&
        (supports.indexOf('sizeByW') !== -1 || supports.indexOf('sizeByWh') !== -1)
      ) {
        return true;
      }
    }
  }

  return false;
}

export function getCustomSizeFromService(service: Service): ImageCandidate[] {
  if (!supportsCustomSizes(service)) {
    return [];
  }

  const imagesSizes: ImageCandidate[] = [];

  // Check for max width in profile.
  const profiles = Array.isArray(service.profile) ? service.profile : [service.profile];

  const pLen = profiles.length;
  for (let x = 0; x < pLen; x++) {
    const profile = profiles[pLen];
    if (typeof profile !== 'string') {
      if (profile.maxHeight || profile.maxWidth) {
        return [
          {
            id: service.id,
            type: 'variable',
            minWidth: 0,
            minHeight: 0,
            maxHeight: (profile.maxHeight || profile.maxWidth) as number,
            maxWidth: (profile.maxWidth || profile.maxHeight) as number,
          },
        ];
      }
    }
  }

  if (service.tiles) {
    const len = service.tiles.length;
    for (let y = 0; y < len; y++) {
      const tile = service.tiles[y];
      if (tile.height || tile.width) {
        imagesSizes.push({
          id: service.id,
          type: 'variable',
          minHeight: 0,
          minWidth: 0,
          maxHeight: tile.height || tile.width,
          maxWidth: tile.width || tile.width,
        });
      }
    }
  }

  return imagesSizes;
}

export function dereferenceImageService(service: Service): Promise<Service> {
  return Promise.resolve(service);
}

export function getSmallestScaleFactorAsSingleImage(service: Service): FixedSizeImageService | null {
  if (!service.width || !service.height) {
    return null;
  }

  if (service.tiles) {
    const len = service.tiles.length;
    for (let i = 0; i < len; i++) {
      const tile = service.tiles[i];
      const targetSize = tile.width || tile.height;
      if (!targetSize) {
        continue;
      }
      const sizeLen = tile.scaleFactors.length;
      const sortedScales = tile.scaleFactors.sort();
      for (let j = 0; j < sizeLen; j++) {
        const size = sortedScales[len - j];
        if (service.width / size < targetSize && service.height / size < targetSize) {
          return {
            id: service.id,
            type: 'fixed-service',
            width: (service.width / size) | 0,
            height: (service.height / size) | 0,
          };
        }
      }
    }
  }
  return null;
}

export function sampledTilesToTiles(width: number, height: number, sampledTiles: ImageTile[]): ImageTile[] {
  const maxDim = width > height ? width : height;
  const len = sampledTiles.length;
  const newTiles: ImageTile[] = [];

  for (let i = 0; i < len; i++) {
    const tile = sampledTiles[i];
    let lastSize = tile.scaleFactors[0];
    let curWidth = maxDim / lastSize;
    const scaleFactors = [lastSize];
    while (curWidth >= tile.width) {
      lastSize = lastSize * 2;
      scaleFactors.push(lastSize);
      curWidth = curWidth / 2;
    }

    newTiles.push({
      ...tile,
      scaleFactors,
    });
  }

  return newTiles;
}

export function isImageService(service: Service): boolean {
  if (!service || !service.profile) {
    return false;
  }

  const profiles = Array.isArray(service.profile) ? service.profile : [service.profile];

  for (const profile of profiles) {
    switch (profile) {
      case STANFORD_IIIF_IMAGE_COMPLIANCE_0:
      case STANFORD_IIIF_IMAGE_COMPLIANCE_1:
      case STANFORD_IIIF_IMAGE_COMPLIANCE_2:
      case STANFORD_IIIF_IMAGE_CONFORMANCE_0:
      case STANFORD_IIIF_IMAGE_CONFORMANCE_1:
      case STANFORD_IIIF_IMAGE_CONFORMANCE_2:
      case STANFORD_IIIF_1_IMAGE_COMPLIANCE_0:
      case STANFORD_IIIF_1_IMAGE_COMPLIANCE_1:
      case STANFORD_IIIF_1_IMAGE_COMPLIANCE_2:
      case STANFORD_IIIF_1_IMAGE_CONFORMANCE_0:
      case STANFORD_IIIF_1_IMAGE_CONFORMANCE_1:
      case STANFORD_IIIF_1_IMAGE_CONFORMANCE_2:
      case IIIF_1_IMAGE_LEVEL_0:
      case IIIF_1_IMAGE_LEVEL_0_PROFILE:
      case IIIF_1_IMAGE_LEVEL_1:
      case IIIF_1_IMAGE_LEVEL_1_PROFILE:
      case IIIF_1_IMAGE_LEVEL_2:
      case IIIF_1_IMAGE_LEVEL_2_PROFILE:
      case IIIF_2_IMAGE_LEVEL_0:
      case IIIF_2_IMAGE_LEVEL_0_PROFILE:
      case IIIF_2_IMAGE_LEVEL_1:
      case IIIF_2_IMAGE_LEVEL_1_PROFILE:
      case IIIF_2_IMAGE_LEVEL_2:
      case IIIF_2_IMAGE_LEVEL_2_PROFILE:
      case IIIF_3_IMAGE_LEVEL_0:
      case IIIF_3_IMAGE_LEVEL_1:
      case IIIF_3_IMAGE_LEVEL_2:
        return true;
    }
  }

  return false;
}
