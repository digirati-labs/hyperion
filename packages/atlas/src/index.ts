export * from './dna';
export * from './layout-engine';
export * from './spacial-content/index';
export * from './temporal-content';
export * from './world';
export * from './world-object';
