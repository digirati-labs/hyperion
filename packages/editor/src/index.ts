import { IIIFBuilder } from './iiif-builder';

export default IIIFBuilder;
