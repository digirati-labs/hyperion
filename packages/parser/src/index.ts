export * from './empty-types';
export * from './normalize';
export * from './traverse';
export * from './serialise';
