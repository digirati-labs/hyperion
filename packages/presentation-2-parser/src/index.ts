export * from './traverse';
export * from './upgrader';
