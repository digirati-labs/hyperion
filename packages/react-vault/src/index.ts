// Context exports.
export * from './context/AnnotationContext';
// export * from './context/AnnotationPageContext';
export * from './context/CanvasContext';
export * from './context/CollectionContext';
export * from './context/ManifestContext';
export * from './context/RangeContext';
export * from './context/VaultContext';
export * from './context/ResourceContext';
export * from './context/VisibleCanvasContext';

// Hook exports
export * from './hooks/useAnnotation';
// export * from './hooks/useAnnotationPage';
export * from './hooks/useAnnotationsAtTime';
export * from './hooks/useCanvas';
export * from './hooks/useCanvasClock';
// export * from './hooks/useCanvasSelector';
// export * from './hooks/useCanvasTimeline';
export * from './hooks/useCollection';
export * from './hooks/useDispatch';
export * from './hooks/useExternalCollection';
export * from './hooks/useExternalManifest';
// export * from './hooks/useImageMatrix';
export * from './hooks/useImageService';
export * from './hooks/useImageTile';
export * from './hooks/useManifest';
// export * from './hooks/useMappedState';
export * from './hooks/usePaintingAnnotations';
export * from './hooks/useRange';
export * from './hooks/useResources';
export * from './hooks/useSearchService';
export * from './hooks/useThumbnail';
export * from './hooks/useVault';
export * from './hooks/useVaultEffect';
export * from './hooks/useVaultSelector';
// export * from './hooks/useVirtualCanvas';

// Viewer exports.
// export * from './viewers/SimpleViewerContext';
// export * from './viewers/SingleCanvasContext';
