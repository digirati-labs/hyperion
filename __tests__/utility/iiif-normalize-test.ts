import { normalize } from '../../src/utility/iiif-normalize';
import { Manifest } from '../../src';

describe('utility/normalize', () => {
  const manifest = (): Manifest => ({
    '@context': ['http://www.w3.org/ns/anno.jsonld', 'http://iiif.io/api/presentation/{{ page.major }}/context.json'],
    id: 'https://example.org/iiif/book1/manifest',
    type: 'Manifest',
    label: { en: ['Image 1'] },
    items: [
      {
        id: 'https://example.org/iiif/book1/canvas/p1',
        type: 'Canvas',
        height: 1800,
        width: 1200,
        items: [
          {
            id: 'https://example.org/iiif/book1/page/p1/1',
            type: 'AnnotationPage',
            items: [
              {
                id: 'https://example.org/iiif/book1/annotation/p0001-image',
                type: 'Annotation',
                motivation: 'painting',
                body: {
                  id: 'http://iiif.io/api/presentation/2.1/example/fixtures/resources/page1-full.png',
                  type: 'Image',
                  format: 'image/png',
                  height: 1800,
                  width: 1200,
                },
                target: 'https://example.org/iiif/book1/canvas/p1',
              },
            ],
          },
        ],
      },
    ],
  });

  test('it can normalize a simple manifest', () => {
    const result = normalize(manifest());

    expect(result.entities).toEqual({
      Annotation: {
        'https://example.org/iiif/book1/annotation/p0001-image': {
          body: {
            id: 'http://iiif.io/api/presentation/2.1/example/fixtures/resources/page1-full.png',
            type: 'ContentResource',
          },
          id: 'https://example.org/iiif/book1/annotation/p0001-image',
          motivation: 'painting',
          target: { id: 'https://example.org/iiif/book1/canvas/p1', type: 'ContentResource' },
          type: 'Annotation',
        },
      },
      AnnotationCollection: {},
      AnnotationPage: {
        'https://example.org/iiif/book1/page/p1/1': {
          id: 'https://example.org/iiif/book1/page/p1/1',
          items: [{ id: 'https://example.org/iiif/book1/annotation/p0001-image', type: 'Annotation' }],
          type: 'AnnotationPage',
        },
      },
      Canvas: {
        'https://example.org/iiif/book1/canvas/p1': {
          height: 1800,
          id: 'https://example.org/iiif/book1/canvas/p1',
          items: [{ id: 'https://example.org/iiif/book1/page/p1/1', type: 'AnnotationPage' }],
          type: 'Canvas',
          width: 1200,
        },
      },
      Collection: {},
      ContentResource: {
        'http://iiif.io/api/presentation/2.1/example/fixtures/resources/page1-full.png': {
          format: 'image/png',
          height: 1800,
          id: 'http://iiif.io/api/presentation/2.1/example/fixtures/resources/page1-full.png',
          type: 'Image',
          width: 1200,
        },
        'https://example.org/iiif/book1/canvas/p1': {
          id: 'https://example.org/iiif/book1/canvas/p1',
          type: 'ContentResource',
        },
      },
      Manifest: {
        'https://example.org/iiif/book1/manifest': {
          '@context': [
            'http://www.w3.org/ns/anno.jsonld',
            'http://iiif.io/api/presentation/{{ page.major }}/context.json',
          ],
          id: 'https://example.org/iiif/book1/manifest',
          items: [{ id: 'https://example.org/iiif/book1/canvas/p1', type: 'Canvas' }],
          label: { en: ['Image 1'] },
          type: 'Manifest',
        },
      },
      Range: {},
      Service: {},
    });

    expect(result.mapping).toEqual({
      'http://iiif.io/api/presentation/2.1/example/fixtures/resources/page1-full.png': 'ContentResource',
      'https://example.org/iiif/book1/annotation/p0001-image': 'Annotation',
      'https://example.org/iiif/book1/canvas/p1': 'Canvas',
      'https://example.org/iiif/book1/manifest': 'Manifest',
      'https://example.org/iiif/book1/page/p1/1': 'AnnotationPage',
    });

    expect(result.resource).toEqual({ id: 'https://example.org/iiif/book1/manifest', type: 'Manifest' });
  });
});
