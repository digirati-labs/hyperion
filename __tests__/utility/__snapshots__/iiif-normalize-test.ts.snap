// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utility/normalize it can normalize a simple manifest 1`] = `
Object {
  "Annotation": Object {
    "https://example.org/iiif/book1/annotation/p0001-image": Object {
      "body": Object {
        "id": "http://iiif.io/api/presentation/2.1/example/fixtures/resources/page1-full.png",
        "type": "ContentResource",
      },
      "id": "https://example.org/iiif/book1/annotation/p0001-image",
      "motivation": "painting",
      "target": Object {
        "id": "https://example.org/iiif/book1/canvas/p1",
        "type": "ContentResource",
      },
      "type": "Annotation",
    },
  },
  "AnnotationCollection": Object {},
  "AnnotationPage": Object {
    "https://example.org/iiif/book1/page/p1/1": Object {
      "behaviour": Array [],
      "homepage": null,
      "id": "https://example.org/iiif/book1/page/p1/1",
      "items": Array [
        Object {
          "id": "https://example.org/iiif/book1/annotation/p0001-image",
          "type": "Annotation",
        },
      ],
      "label": null,
      "logo": Array [],
      "metadata": Array [],
      "motivation": null,
      "rendering": Array [],
      "requiredStatement": null,
      "rights": null,
      "seeAlso": Array [],
      "service": Array [],
      "summary": null,
      "thumbnail": Array [],
      "type": "AnnotationPage",
    },
  },
  "Canvas": Object {
    "https://example.org/iiif/book1/canvas/p1": Object {
      "annotations": Array [],
      "behaviour": Array [],
      "duration": 0,
      "height": 1800,
      "homepage": null,
      "id": "https://example.org/iiif/book1/canvas/p1",
      "items": Array [
        Object {
          "id": "https://example.org/iiif/book1/page/p1/1",
          "type": "AnnotationPage",
        },
      ],
      "label": null,
      "logo": Array [],
      "metadata": Array [],
      "motivation": null,
      "navDate": null,
      "partOf": Array [],
      "posterCanvas": null,
      "rendering": Array [],
      "requiredStatement": null,
      "rights": null,
      "seeAlso": Array [],
      "service": Array [],
      "summary": null,
      "thumbnail": Array [],
      "type": "Canvas",
      "width": 1200,
    },
  },
  "Collection": Object {},
  "ContentResource": Object {
    "http://iiif.io/api/presentation/2.1/example/fixtures/resources/page1-full.png": Object {
      "format": "image/png",
      "height": 1800,
      "id": "http://iiif.io/api/presentation/2.1/example/fixtures/resources/page1-full.png",
      "type": "Image",
      "width": 1200,
    },
    "http://myhomepage.com": Object {
      "id": "http://myhomepage.com",
      "type": "ContentResource",
    },
    "https://example.org/iiif/book1/canvas/p1": Object {
      "id": "https://example.org/iiif/book1/canvas/p1",
      "type": "ContentResource",
    },
  },
  "Manifest": Object {
    "https://example.org/iiif/book1/manifest": Object {
      "@context": Array [
        "http://www.w3.org/ns/anno.jsonld",
        "http://iiif.io/api/presentation/{{ page.major }}/context.json",
      ],
      "annotations": Array [],
      "behaviour": Array [],
      "homepage": Object {
        "id": "http://myhomepage.com",
        "type": "ContentResource",
      },
      "id": "https://example.org/iiif/book1/manifest",
      "items": Array [
        Object {
          "id": "https://example.org/iiif/book1/canvas/p1",
          "type": "Canvas",
        },
      ],
      "label": Object {
        "en": Array [
          "Image 1",
        ],
      },
      "logo": Array [],
      "metadata": Array [],
      "motivation": null,
      "navDate": null,
      "partOf": Array [],
      "posterCanvas": null,
      "rendering": Array [],
      "requiredStatement": null,
      "rights": null,
      "seeAlso": Array [],
      "service": Array [],
      "start": null,
      "structures": Array [],
      "summary": null,
      "thumbnail": Array [],
      "type": "Manifest",
      "viewingDirection": "left-to-right",
    },
  },
  "Range": Object {},
  "Selector": Object {},
  "Service": Object {},
}
`;
